db/schema_applier.cc

```diff
@@ -841,6 +842,79 @@ future<> schema_applier::prepare(utils::chunked_vector<mutation>& muts) {
     }
 }
 
+class pending_schema_getter : public service::schema_getter {
+private:
+    schema_applier& _sa;
+    distributed<replica::database>& _db;
+
+public:
+    pending_schema_getter(schema_applier& sa) :
+            _sa(sa), _db(sa._proxy.local().get_db()) {
+    };
+
+    virtual replica::column_family* find_column_family(const table_id& uuid) const override {
+        auto& tables_and_views = _sa._affected_tables_and_views.tables_and_views.local();
+        for (auto& schema : tables_and_views.tables.dropped) {
+            if (schema->id() == uuid) {
+                return nullptr;
+            }
+        }
+        for (auto& schema : tables_and_views.views.dropped) {
+            if (schema->id() == uuid) {
+                return nullptr;
+            }
+        }
+        auto cf_it = _sa._affected_tables_and_views.table_shards.find(uuid);
+        if (cf_it != _sa._affected_tables_and_views.table_shards.end()) {
+            return &(*cf_it->second);
+        }
+        return &_db.local().find_column_family(uuid);
+    };
+
+    virtual flat_hash_map<sstring, replica::keyspace*> get_keyspaces() const override {
+        flat_hash_map<sstring, replica::keyspace*> out;
+        for (auto& [name, ks] : _db.local().get_keyspaces()) {
+            out.emplace(name, &ks);
+        }
+        for (const auto& name : _sa._affected_keyspaces.names.dropped) {
+            out.erase(name);
+        }
+        for (const auto& ks_all_shards : _sa._affected_keyspaces.created) {
+            auto& ks = *ks_all_shards[this_shard_id()];
+            auto name = ks.metadata()->name();
+            out.insert_or_assign(name, &ks);
+        }
+        // Note that this may return stale information for altered keyspaces
+        // but we can't easily return replica::keyspace* in such case because
+        // keyspace_change doesn't reference it (and it can't since it's not updated yet).
+        // So we'd need to build extra indirection to manipulate erms and replication strategies
+        // in such special case. But we simplify here as this stale erm information will be
+        // corrected later by committing keyspace changes (see comment in commit_on_shard()) and
+        // replication strategy can't currently change to per_table after keyspace is created.
+        return out;
+    };
+
+    virtual future<> for_each_table_gently(std::function<future<>(table_id, lw_shared_ptr<replica::table>)> f) const override {
+        auto ff = [this, &f](table_id id, lw_shared_ptr<replica::table> t) -> future<> {
+            if (find_column_family(id)) {
+                  co_await f(id, t);
+            }
+        };
+        return _db.local().get_tables_metadata().for_each_table_gently(ff);
```

Comment:
What about new tables?

Reply:
but it's problematic because at this point there is no table object yet, and because committing tables uses _pending_token_metadata which is up to date, the new change is included there.